equinox/_serialisation.py

```diff
@@ -105,6 +116,14 @@ def default_deserialise_filter_spec(f: BinaryIO, x: Any) -> Any:
         return np.load(f)
     elif isinstance(x, (bool, float, complex, int)):
         return np.load(f).item()
+    elif isinstance(x, np.generic):
+        y = np.load(f)
+        try:
+            # `item()` converts back to base types like `float`
+            # so we have to use the constructor
+            return type(x)(y.astype(x.dtype).item())
+        except ValueError:
+            return type(x)(jnp.frombuffer(y, dtype=x.dtype)[0].item())
```

Comment:
What's going on here? This seems like a lot. What goes wrong with e.g. `type(x)(y)`?

Reply:
Yeah, it's a bit unfortunate.

The immediate error if you just do `type(x)(y)` is: `No cast function available`.

My understanding: The issue is that `type(y.item())` is just `bytes` because when you `np.load` a `bfloat16` (and presumably other types numpy doesn't know about) it just treats them as bytes. So we have to turn the bytes into a JAX array and then interpret them as the proper dtype. `frombuffer` is the only function I know of for converting bytes into an array, but it "Interpret[s] a buffer as a 1-dimensional array." so we then have to `[0]` to turn that into a 0D array. 

One alternative is:

```py
pos = f.tell()
try:
  type(x)(np.load(f).astype(x.dtype).item())
except ValueError:
  f.seek(pos)
  type(x)(jnp.load(f).astype(x.dtype).item())
```
which is a little more straightforward in some ways, but also involves more IO so is presumably slower.

Reply:
I don't think this branch should trigger for `bfloat16`:
```python
import jax.numpy as jnp

x = jnp.bfloat16(1)
print(isinstance(x, jnp.generic))  # False
print(type(x))  # It's a JAX array
```
That is, I think this case should already be handled by the `isinstance(x, jax.Array)` path.

Reply:
Ah, this is is for handling scalars directly (admittedly probably not many bfloat16 scalars but perhaps makes more sense for np.float32 and whatever other `np.generic` scalars people may want to use):
```py
>>> jnp.bfloat16(1)
Array(1, dtype=bfloat16)
>>> jax.dtypes.bfloat16(1)
1
>>> isinstance(jax.dtypes.bfloat16(1), np.generic)
True
```

Reply:
(And I checked that the branch is indeed triggered. The test does not pass unless I add this branch.)

Reply:
Gotcha; I see the distinction between `jax.{numpy,dtypes}.bfloat16`.
So I've just checked, and `jnp.load` will correctly deserialise a `jax.dtypes.bfloat16`. As such I suspect the correct thing to do here is to have both ser+de run `if is_array_like(x): jnp.{save, load}`. Note that `is_array_like(x) == isinstance(x, (bool, int, float, complex, np.generic, np.ndarray, jax.Array))`, so this is just smushing multiple branches together into a single implementation.